mohawk/util.py

```diff
@@ -268,3 +275,110 @@ def normalize_header_attr(val):
+    # on (although again, the canonical implementation does not do this)
+    clientId = resource.credentials['id'].replace("\\", "")
+
+    # b64encode works only with bytes in python3, but all of our parameters are
+    # in unicode, so we need to encode them. The cleanest way to do this that
+    # works in both python 2 and 3 is to use string formatting to get a
+    # unicode string, and then explicitly encode it to bytes.
+    inner_bewit = u"{id}\\{exp}\\{mac}\\{ext}".format(
+        id=clientId,
+        exp=resource.timestamp,
+        mac=mac,
+        ext=ext,
+    )
+    inner_bewit_bytes = inner_bewit.encode('ascii')
+    bewit_bytes = urlsafe_b64encode(inner_bewit_bytes)
+    # Now decode the resulting bytes back to a unicode string
+    return bewit_bytes.decode('ascii')
+
+
+bewittuple = namedtuple('bewittuple', 'id exp mac ext')
+
+def parse_bewit(bewit):
+    """Returns the parts of the bewit as a namedtuple:
+        (id, exp, mac, ext)
+    Input it base64-encoded bewit string
+    """
+    decoded_bewit = b64decode(bewit).decode('ascii')
+    return bewittuple(*decoded_bewit.split("\\", 3))
+
+def strip_bewit(url):
+    """Strips the bewit parameter out of the url.
+    Returns (encoded_bewit, stripped_url)
+    Raises ValueError if no bewit found."""
+    m = re.search('[?&]bewit=([^&]+)', url)
+    if not m:
+        raise ValueError('no bewit found')
+    bewit = m.group(1)
+    stripped_url = url[:m.start()] + url[m.end():]
+    return bewit, stripped_url
+
+
+def check_bewit(url, credentials_map, timestamp=None, nonce=''):
+    """Returns True if the resource has a valid bewit parameter attached; False otherwise.
+    If timestamp is None, then the current time is used."""
+
+    # This import is here to avoid circular imports
+    # base -> util -> base
+    # ugh
+    from .base import Resource
+    raw_bewit, stripped_url = strip_bewit(url)
+    bewit = parse_bewit(raw_bewit)
+    if bewit.id not in credentials_map:
+        raise CredentialsLookupError('Could not find credentials for ID {0}'
+                                     .format(bewit.id))
+
+    # Check that the timestamp isn't expired
+    if timestamp is None:
+        timestamp = time.time()
+    if int(bewit.exp) < timestamp:
+        raise TokenExpired('bewit with UTC timestamp {ts} has expired; '
+                           'it was compared to {now}'
+                           .format(ts=bewit.exp, now=timestamp),
+                           localtime_in_seconds=timestamp,
+                           www_authenticate=''
+                           )
+
+
+    res = Resource(url=stripped_url,
+                   method='GET',
+                   credentials=credentials_map[bewit.id],
+                   timestamp=bewit.exp,
+                   nonce='',
+                   ext=bewit.ext,
+                   )
+    mac = calculate_mac('bewit', res, None)
+    mac = mac.decode('ascii')
+
+    if mac != bewit.mac:
```

Comment:
you should check the mac before checking the timestamp because checking the mac means you can trust that the timestamp value has not been tampered with.